components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Activity.kt

```diff
@@ -5,50 +5,86 @@
 package mozilla.components.support.ktx.android.view
 
 import android.app.Activity
+import android.os.Build
+import android.view.View
 import android.view.ViewTreeObserver
+import android.view.WindowInsets.Type.statusBars
 import android.view.WindowManager
 import androidx.annotation.VisibleForTesting
 import androidx.core.view.WindowInsetsCompat
 import androidx.core.view.WindowInsetsControllerCompat
 import mozilla.components.support.base.log.logger.Logger
 
 /**
- * Attempts to call immersive mode using the View to hide the status bar and navigation buttons.
- * @param onWindowFocusChangeListener optional callback to ensure immersive mode is stable
- * Note that the callback reference should be kept by the caller and be used for [exitImmersiveModeIfNeeded] call.
+ * Attempts to enter immersive mode - fullscreen with the status bar and navigation buttons hidden.
+ * This will automatically register and use an
+ * - an inset listener: [View.OnApplyWindowInsetsListener] on API 30+ or
+ * [View.OnSystemUiVisibilityChangeListener] for below APIs
+ * - a focus listener: [ViewTreeObserver.OnWindowFocusChangeListener]
+ *
+ * to restore immersive mode if interactions with various other widgets like the keyboard or dialogs
+ * got the activity out of immersive mode without [exitImmersiveModeIfNeeded] being called.
  */
-fun Activity.enterToImmersiveMode(
-    onWindowFocusChangeListener: ViewTreeObserver.OnWindowFocusChangeListener? = null
-) {
+fun Activity.enterToImmersiveMode() {
+    setAsImmersive()
+    enableImmersiveModeRestore()
+}
+
+@VisibleForTesting
+internal fun Activity.setAsImmersive() {
     window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
     window.getWindowInsetsController().apply {
         hide(WindowInsetsCompat.Type.systemBars())
         systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
     }
+}
+
+/**
+ * Anytime a new Window is focused like of a Dialog or of the keyboard the activity
+ * will exit immersive mode.
+ * This will observe such events and set again the activity to be immersive
+ * the next time it gets focused.
+ */
+@VisibleForTesting
+internal fun Activity.enableImmersiveModeRestore() {
+    window.decorView.viewTreeObserver?.addOnWindowFocusChangeListener(onWindowFocusChangeListener)
 
-    // We need to make sure system bars do not become permanently visible after interactions with content
-    // see https://github.com/mozilla-mobile/fenix/issues/20240
-    onWindowFocusChangeListener?.let {
-        window.decorView.viewTreeObserver?.addOnWindowFocusChangeListener(it)
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+        window.decorView.setOnApplyWindowInsetsListener { view, insets ->
+            println(view)
+            if (insets.isVisible(statusBars())) {
+                setAsImmersive()
+            }
+            insets
```

Comment:
nit: Do we need `insets` here? As it doesn't looks as we are using it.

Reply:
It's the return of windowInsetsListener.
The `onApplyWindowInsets` method must return `WindowInsets` - https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:core/core/src/main/java/androidx/core/view/OnApplyWindowInsetsListener.java;l=45;drc=cb04420e742501081487e3b18851d5674dc07536